test: serialize and deserialize RPC error messages

Alizter opened the pull request:
We demonstrate a bug with how error messages are serialized. Currently there are two "Error:" prefixes after serializing and deserializing. This is because the prefix hasn't been stripped in the initial serialization. We should strip the prefix since the severity information is recorded elsewhere in the diagnostic data.

Alizter committed:
test: serialize and deserialize RPC error messages

We demonstrate a bug with how error messages are serialized. Currently
there are two "Error:" prefixes after serializing and deserializing.
This is because the prefix hasn't been stripped in the initial
serialization. We should strip the prefix since the severity information
is recorded elsewhere in the diagnostic data.

Signed-off-by: Ali Caglayan <alizter@gmail.com>

Alizter commented:
@rgrinberg I think the correct fix for this would be to make `user_message.ml` remember prefixes in the type. Otherwise, I don't see an easy way of removing `Error:` short of regexing.

rgrinberg commented on src/dune_engine/build_system.mli:
```diff
@@ -108,6 +109,14 @@ module Error : sig
+    :  id:Id.t
+    -> description:[ `Exn of Exn_with_backtrace.t | `Diagnostic of Compound_user_error.t ]
+    -> ?dir:Path.t
+    -> ?promotion:Diff_promotion.Annot.t
```
Don't use optional arguments without a concrete reason.

Also, this function should be moved to a `For_tests` submodule.

rgrinberg commented on src/dune_rpc_impl/dune_rpc_impl.ml:
```diff
@@ -5,6 +5,7 @@ module For_handlers = For_handlers
 module Private = Dune_rpc_client.Private
 module Watch_mode_config = Watch_mode_config
 module Where = Dune_rpc_client.Where
+module Diagnostics = Diagnostics
```
Since you only need one function, you can just export the function you need. Also, it should be accessible as `Diagnostics.For_tests.xxx`.

Alizter commented:
@rgrinberg I've pushed another test which shows how location printing is kind of funny. I need some help scrubbing the output however.

Alizter commented on test/expect-tests/dune_rpc/dune:
```diff
@@ -5,10 +5,13 @@
   dune_rpc_impl
   dune_rpc_client
+  dune_engine
+  dune_re
```
Need to remove this, didn't use it in the end.

Alizter commented on test/expect-tests/dune_rpc/dune_rpc_tests.ml:
```diff
@@ -925,3 +925,37 @@ let%expect_test "print digests for all public RPCs" =
+    File "Bar", line 1, characters 2-3:
+    Error: An error with location!
+    ------- RPC ------
+    Error: (In directory /Foo)
+    File "/Foo/Bar", line 1, characters 2-3:
+
+    Error: An error with location! |}]
```
@rgrinberg It seems to me like we shouldn't have to print the `In directory` part as the loc already contains this info. WDYT?

rgrinberg replied:
Yeah, I guess it's redundant if the location is already there.

rgrinberg commented on test/expect-tests/dune_rpc/dune_rpc_tests.ml:
```diff
@@ -925,3 +925,37 @@ let%expect_test "print digests for all public RPCs" =
+  Stdune.Path.set_root (Stdune.Path.External.of_filename_relative_to_initial_cwd ".")
+;;
+
+let test ?dir ?promotion main =
```
no optional arguments

Alizter commented on otherlibs/stdune/src/stdune.ml:
```diff
@@ -30,6 +30,32 @@ module Pp = struct
     Ordering.of_int (Pp.compare (fun a b -> Ordering.to_int (compare a b)) x y)
   ;;
+
+  let to_dyn tag_to_dyn t =
```
I've added Pp.to_dyn since I've been wanting this for a while and it make our life easier when testing.

Alizter commented:
@rgrinberg I've also included dyn serialisation so we can more finely compare the Pp structure that has been serialised.

rgrinberg commented on test/expect-tests/dune_rpc/dune:
```diff
@@ -5,10 +5,12 @@
   dune_rpc_server
   dune_rpc_impl
   dune_rpc_client
+  dune_engine
```
can you separate this into a different test library? Something like `dune_rpc_impl_tests`.

This library is specifically for testing the RPC machinery and not concrete dune rpc things

Alizter replied:
Sure

Alizter commented:
@rgrinberg I've moved the tests now and added yet another example of an error message that changes. All these examples will be fixed by my upcoming PR.

rgrinberg commented on src/dune_engine/build_system.mli:
```diff
@@ -88,6 +88,7 @@ module Error : sig
 
     module Map : Map.S with type key = t
 
+    val gen : unit -> t
```
Do you need to make this function public? Can you just make `make` generate a new id? It's only used for the tests anyway.

Alizter replied:
Good point

Alizter committed:
Merge branch 'main' into ps/branch/test__serialize_and_reserialize_rpc_error_messages

Alizter committed:
Merge branch 'main' into ps/branch/test__serialize_and_reserialize_rpc_error_messages
